src/alignment/poa.rs

```diff
@@ -0,0 +1,529 @@
+// Copyright 2017-2018 Brett Bowman, Jeff Knaggs
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Partial-Order Alignment for fast alignment and consensus of multiple homologous sequences.
+//!
+//! For the original concept and theory, see:
+//! * Lee, Christopher, Catherine Grasso, and Mark F. Sharlow. "Multiple sequence alignment using
+//! partial order graphs." Bioinformatics 18.3 (2002): 452-464.
+//! * Lee, Christopher. "Generating consensus sequences from partial order multiple sequence
+//! alignment graphs." Bioinformatics 19.8 (2003): 999-1008.
+//!
+//! For a modern reference implementation, see poapy:
+//! https://github.com/ljdursi/poapy
+//!
+//! # Example
+//!
+//! ```
+//! use bio::alignment::poa::*;
+//! let x = b"AAAAAAA";
+//! let y = b"AABBBAA";
+//! let z = b"AABCBAA";
+//!
+//! let score = |a: u8, b: u8| if a == b { 1i32 } else { -1i32 };
+//! let mut aligner = Aligner::from_string(x, -1, &score);
+//! assert_eq!(aligner.global(z).score, 1);
+//! aligner.add_sequence(y);
+//! // z differs from its closest homologue in the graph by one base
+//! assert_eq!(aligner.global(z).score, 5);
```

Comment:
I pushed a new method for Aligner that allows initialization without a sequence. Calling `global(...)` on an empty Aligner is an error.

In that example would global, local, and semi-global alignments have to be computed and stored in the Aligner struct at the time of add_sequence? So that the `global()` call is just accessing some global_alignment member?

I want to be able to align without adding the sequence to the graph. For example, a user could construct a POA from a bunch of semi-globally aligned contigs (eg. from de novo assembly) and then score the alignments of different reference sequences locally without mutating the POA.

Reply:
So to clarify would
```rust
aligner.add_sequence(x);
aligner.add_sequence(y);
aligner.add_sequence(z);

aligner_y.add_sequence(x);
aligner_y.add_sequence(y);

assert_eq!(aligner.global(),  aligner_y.global(z));
```
i.e. `global()` retrieves the most recent alignment?

To add a sequence we have to align it first, so if we call `global()` to retrieve the most recent alignment we will either have to store it or recompute it.

I've pushed an example that avoids redundant alignment like this:
```rust
let alignment = aligner.global(y);
aligner.add_alignment(alignment, y);
aligner.global(z)
```
(with the ugliness of passing `y` at two locations that could be solved by storing the query sequence in the `Alignment` struct.) Incidentally, I would prefer being able to write `aligner.add_alignment(aligner.global(y), y)` but with the traceback matrix stored as a member of Aligner the borrowing gets complicated. One solution would be to move away from storing so much state in the aligner instance.

For the local and semi-global implementations my plan is to follow the pairwise module's design. So generalizing the `global` method's implementation into a `custom` method which accepts parameters for soft clipping that alter the locality.

Reply:
Mhm, maybe I misunderstood this. I always though a partial-order-alignment yields a multiple sequence alignment (e.g. in the form of a graph, or as a matrix of alignments). However, here, you always only get a single alignment of one sequence right?

Reply:
yeah, the partially ordered graph is an aggregation of these alignments between itself and regular sequences. So the sequences of `AlignmentOperations` are still the classic indels and mismatches from pairwise alignment and cannot describe the structure of the entire graph.

We can interpret the `Aligner.graph` member as an MSA without further processing. In the issue thread bnbowman used petgraph's dot file output and I quite like this presentation too. I haven't put any thought into outputting the graph in a clustal style format but am now realizing that this might be an important feature.. Another operation on the graph datastructure would be flattening it into a consensus sequence.

so the alignments I'm referring to above are just the atomic operations on the graph that build it and permit using it for the narrow set of analyses that I'm interested in.

Reply:
Ok. Would the following make sense?

```
let mut aligner = Aligner::new(-1, &score);
aligner.global(x);
let z_aln = aligner.global(z);
assert_eq!(z_aln.score, 1);
let y_aln = aligner.global(y);
assert_eq!(y_aln.score, 5);
let s_local_aln = aligner.local(s);
let t_semiglobal = aligner.semiglobal(t);

let msa = aligner.graph();
```

Reply:
The different alignment methods directly return the alignment and at the same time update the graph.

Reply:
Alternatively:

```
let mut aligner = Aligner::new(-1, &score);
aligner.global(x).add_to_graph();
// align and add to graph
let z_aln = aligner.global(z).add_to_graph().alignment();
// align and don't add to graph
let y_aln = aligner.global(y).alignment();
```

Here, both the addition to the graph and the returning of the alignment is optional.

Reply:
I have seen your comment about the recycling. Still thinking that above proposal would be the better API though. Seems more streamlined, also it is unclear from the name what kind of alignment `add_sequence` performs. By explicitly invoking the type of alignment in order to add something to the graph, this problem does not occur.

Reply:
I really like the method chaining syntax but I can not find a way to make it work without making the structs more mutable than they have to be. If `aligner.global(x)` returns an `Alignment` then that struct needs to carry a mutable reference to the `Aligner` that created it to implement `add_to_graph()`. Otherwise, if `global(x)` mutates itself and stores the most recent Alignment then the ownership get complicated because each method in the chain mutates the same Self.

I have been experimenting with making things less mutable by moving the Traceback member from Aligner altogether because it only needs to exist during the alignment step. This is what I gather from these guidelines: https://rust-lang-nursery.github.io/api-guidelines/flexibility.html

I think a solution that satisfies both ideals may be the builder pattern, https://doc.rust-lang.org/1.0.0/style/ownership/builders.html: The method chaining could be implemented for AlignerBuilders but Aligners and Alignments get to stay immutable.

I totally agree about `add_sequence`.